005-graphql-api.md

```diff
@@ -0,0 +1,153 @@
+  * https://github.com/appwrite/appwrite/issues/511
+
+## Summary
+
+[summary]: #summary
+
+<!-- Brief explanation of the proposed contribution. Write your answer below. -->
+
+GraphQL is a query language for APIs and a runtime for fulfilling those queries with your existing data. GraphQL provides a complete and understandable description of the data in your API, gives clients the power to ask for exactly what they need and nothing more, makes it easier to evolve APIs over time, and enables powerful developer tools.
+
+### GraphQL vs REST API
+
+A REST API is an architectural concept for network-based software. GraphQL, on the other hand, is a query language, a specification, and a set of tools that operates over a single endpoint using HTTP. In addition, over the last few years, REST has been used to make new APIs, while the focus of GraphQL has been to optimize for performance and flexibility.
+
+While Appwrite REST API is an easy way to get started with Appwrite quickly, GraphQL will allow better performance for specific use-cases and new ways to integrate Appwrite with platforms that we don't yet support that relies on that protocol. Ex: iOS, Android, or Gatsby.js.
+
+The new Appwrite GraphQL API is not designed or aimed to replace the REST API, but rather to complete it and work well alongside it.
+
+## Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+**What problem are you trying to solve?**
+
+<!-- Write your answer below. -->
+
+While Appwrite REST API is easy and simple to use, a lot of developers would prefer to integrate their API and client applications using a GraphQL API. GraphQL can be a bit more complex but also more flexible and introduce better performance in some use-cases.
+
+**What is the context or background in which this problem exists?**
+
+<!-- Write your answer below. -->
+
+Developers using Appwrite should not be forced to integrate with the server using a specific protocol. Appwrite should be protocol agnostic and provide multiple integrations API for different use-cases and ease of use for developers who come with particular perforations or existing knowledge.
+
+**Once the proposal is implemented, how will the system change?**
+
+<!-- Write your answer below. -->
+
+We'll introduce a new API route called POST /v1/graphql. The new endpoint will **parse**, and **route** GraphQL requests to existing Appwrite controllers and routes and return the same response models as the Appwrite REST API endpoints.
+
+<!-- Please avoid discussing your proposed solution. -->
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+### Create a New Route
+
+Create a new route called POST /v1/graphql. The new endpoint will initialize the GraphQL parser [webonyx/graphql-php](https://github.com/webonyx/graphql-php). We should use Utopia PHP routes object callbacks and metadata to init GraphQL queries and mutations.
+
+### Get Route by Method Name
```

Comment:
I don't fully understand what this means. Does it mean that we will be using the same existing routes as GraphQL API actions? I am not sure what it means practically.

Reply:
It means we'll use the same controllers we use in the REST API as the execution targets for any GraphQL queries we get from end users.